volmigrate/migration.go

```diff
@@ -0,0 +1,39 @@
+package volmigrate
+
+import (
+	"fmt"
+
+	"github.com/contiv/errored"
+	"github.com/contiv/volplugin/volmigrate/backend"
+)
+
+// Migrations is used to sort by the version number.
+type Migrations []*Migration
+
+func (m Migrations) Len() int           { return len(m) }
+func (m Migrations) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+func (m Migrations) Less(i, j int) bool { return m[i].Version < m[j].Version }
+
+// Migration represents a migration that volmigrate can run.
+type Migration struct {
+	Version     int64  // each new migration increments this value, must be >= 1
+	Description string // e.g., "volplugin 1.3 release"
+	runner      func(backend.Backend) error
+}
+
+// Run executes the migration using a supplied backend.Backend
+func (m *Migration) Run(b backend.Backend) error {
+	fmt.Printf("Running migration #%d against %s backend\n", m.Version, b.Name())
+
+	if err := m.runner(b); err != nil {
+		return errored.Errorf("Encountered error during migration %d", m.Version).Combine(err)
+	}
+
+	if err := b.UpdateSchemaVersion(m.Version); err != nil {
+		return errored.Errorf("Successfully applied migration but failed to update schema version key").Combine(err)
```

Comment:
Without transaction support, there's nothing we can do except print a warning and have the end user rectify the situation manually.

In an optimal world, the entire migration + the schema version key update would be a single atomic unit of work.  My original design did include `BeginTransaction()` and `CommitTransaction()` as part of the Backend interface, but since etcd2 doesn't support transactions, they were just a NOOP.  If we want to support transactions in the future, it's trivial to add them (add interface functions + wrap `m.Run()` in migration.go).